sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -200,214 +165,233 @@ export class EventHubReceiver extends LinkEntity {
+        "[%s] Since the user did not close the receiver and the session error is not " +
+        "retryable, we let the user know about it by calling the user's error handler.",
+        this._context.connectionId
+      );
+      this._onError(error);
+    }
+  }
+
+  private async _onAmqpSessionClose(context: EventContext): Promise<void> {
+    const amqpReceiver = this._receiver || context.receiver;
+    const sessionError = context.session && context.session.error;
+
+    if (sessionError) {
+      log.error(
+        "[%s] 'session_close' event occurred for receiver '%s' with address '%s'. " +
+        "The associated error is: %O",
+        this._context.connectionId,
+        this.name,
+        this.address,
+        sessionError
+      );
+    }
+
+    if (!amqpReceiver || amqpReceiver.isSessionItselfClosed()) {
+      log.error(
+        "[%s] 'session_close' event occurred on the session of receiver '%s' with " +
+        "address '%s' and the sdk did not initiate this. Moreover the receiver is already " +
+        "re-connecting. Hence not calling detached from the _onAmqpSessionClose() handler.",
+        this._context.connectionId,
+        this.name,
+        this.address
+      );
+      return;
+    }
+
+    if (!this.isConnecting) {
+      log.error(
+        "[%s] 'session_close' event occurred on the session of receiver '%s' with " +
+        "address '%s' and the sdk did not initiate this. Hence calling detached from the " +
+        "_onAmqpSessionClose() handler.",
+        this._context.connectionId,
+        this.name,
+        this.address
+      );
+      // session close event not initiated by SDK, so calling detached handler.
+      await this.onDetached(sessionError);
+    } else {
+      log.error(
+        "[%s] 'session_close' event occurred on the session of receiver '%s' with " +
+        "address '%s' and the sdk did not initiate this. Moreover the receiver is already " +
+        "re-connecting. Hence not calling detached from the _onAmqpSessionClose() handler.",
+        this._context.connectionId,
+        this.name,
+        this.address
+      );
+    }
+  }
+
+  private async _onAmqpClose(context: EventContext): Promise<void> {
+    const amqpReceiver = this._receiver || context.receiver;
+    if (!amqpReceiver || amqpReceiver.isItselfClosed()) {
+      log.error(
+        "[%s] 'receiver_close' event occurred on the receiver '%s' with address '%s' " +
+        "because the sdk initiated it. Hence not calling detached from the _onAmqpClose" +
+        "() handler.",
+        this._context.connectionId,
+        this.name,
+        this.address
+      );
+      return;
+    }
+
+    const amqpError = amqpReceiver.error;
+    if (amqpError) {
+      log.error(
+        "[%s] 'receiver_close' event occurred for receiver '%s' with address '%s'. " +
+        "The associated error is: %O",
+        this._context.connectionId,
+        this.name,
+        this.address,
+        amqpError
+      );
+    }
```

Comment:
The order of "if closed, then return" followed by "log error if exists" is flipped in `_onAmqpSessionClose`